File: Example/Ex8_poincare_map.m
% poincare map at x = 1 - mu

%% Start of script
close all;  %close all figures
clear;      %clear all variables
clc;        %clear the command terminal
format long
%warning off

% line width
set(0,'DefaultLineLineWidth',1.5) % default 0.5pt
set(0,'DefaultAxesLineWidth',1.5)
set(0,'DefaultTextLineWidth',1.5)

% font size
set(0,'DefaultTextFontSize',20)
set(0,'DefaultAxesFontSize',20)

% font name
set(0,'DefaultTextFontName','Times New Roman')
set(0,'DefaultAxesFontName','Times New Roman')
set(0,'DefaultTextInterpreter','Latex')
set(0,'DefaultLegendInterpreter','Latex')

% figure color
set(0,'DefaultFigureWindowStyle','docked');
set(gcf,'Color','none');
set(gca,'Color','none');
set(gcf,'InvertHardCopy', 'off');

close

current_pass = pwd;
addpath(replace(current_pass, 'Examples', 'Functions'));

myTimer = tic;        %start timer


%% initial settings
[mu,~,~,~] = parameter(6); % Sun-Jupiter
[L1,L2,L3,L4,L5] = librationPoints(mu);

options_ODE = odeset('RelTol',3e-14, 'AbsTol',1e-14);
options_ODE_poincare = odeset('RelTol',3e-14, 'AbsTol',1e-14,'Events',@(t,x) fun_odestop_CR3BP_Body2_vertical(t,x,mu) );



%% differential correction
iteration_DC_max = 100;
threshold = 1e-12;

% L1 Lyapunov orbit
x0_1 = [0.926672517344211 0 0  0 0.045684284381193 0]';
t0_1 = 2.915640890603734/2;
for iteration = 1:iteration_DC_max
  [x_n_1,t_n_1,C_L1] = fun_differential_correction_cr3bp(x0_1,t0_1,mu,options_ODE);

  tspan = [0 2*t_n_1];
  [t_corrected, x_corrected] = ode113(@(t,x) fun_cr3bp(t,x,mu), tspan, x_n_1, options_ODE);

  x_error = norm(x_corrected(end,:) - x_corrected(1,:));
  %disp( strcat('x_error = ',num2str(x_error)) );
  if x_error < threshold
    break;
  end

  if x_error > 1e+3
    disp('calculation diverged');
    return;
  end

  if iteration == iteration_DC_max
    disp('do not finish');
    return;
  end

  x0_1 = x_n_1;
  t0_1 = t_n_1;
end
x_L1 = x_corrected; % C_L1 = 3.036999999999999
t_L1 = t_n_1;

% L2 Lyapunov orbit
x0_2 = [1.072915577273646 0 0 0 -0.025945940908770 0]';
t0_2 = 3.184772418214781/2;
for iteration = 1:iteration_DC_max
  [x_n_2,t_n_2,C_L2] = fun_differential_correction_cr3bp(x0_2,t0_2,mu,options_ODE);

  tspan = [0 2*t_n_2];
  [t_corrected, x_corrected] = ode113(@(t,x) fun_cr3bp(t,x,mu), tspan, x_n_2, options_ODE);

  x_error = norm(x_corrected(end,:) - x_corrected(1,:));
  %disp( strcat('x_error = ',num2str(x_error)) );
  if x_error < threshold
    break;
  end

  if x_error > 1e+3
    disp('calculation diverged');
    return;
  end

  if iteration == iteration_DC_max
    disp('do not finish');
    return;
  end

  x0_2 = x_n_2;
  t0_2 = t_n_2;
end
x_L2 = x_corrected; % C_L2 = 3.036969030648278
t_L2 = t_n_2;



%% zero velocity curve
[x_curve,y_curve] = meshgrid(0.9:1e-4:1.1,-0.1:1e-4:0.1);
[a,b] = size(x_curve);
z_curve = zeros(a,b);

x_curve = reshape(x_curve, [a*b,1]);
y_curve = reshape(y_curve, [a*b,1]);
z_curve = reshape(z_curve, [a*b,1]);

r1 = sqrt((x_curve+mu).^2+y_curve.^2+z_curve.^2);
r2 = sqrt((x_curve-(1-mu)).^2+y_curve.^2+z_curve.^2);
U = 1/2.*(x_curve.^2+y_curve.^2) + (1-mu)./r1 + mu./r2;
C = 2.*U;

x_curve = reshape(x_curve, [a,b]);
y_curve = reshape(y_curve, [a,b]);
C = reshape(C, [a,b]);



%% manifold
tspan_u = [0 9];
tspan_s = [10 0];
xpert = 1e-10;
N = 40;

% L1
[~,~,~, XU_right,~] = fun_manifold_cr3bp(mu, x_L1(1,:)', 2*t_L1, N, xpert, options_ODE);

% L2
[XS_left,~,~,~,~] = fun_manifold_cr3bp(mu, x_L2(1,:)', 2*t_L2, N, xpert, options_ODE);



%% poincare map
yu_right_f1 = zeros(N,6);
yu_right_f2 = NaN(N,6);
ys_left_f1 = zeros(N,6);
ys_left_f2 = NaN(N,6);

f1 = figure();
hold on
contourf(x_curve, y_curve, -C, [-C_L1 -C_L1],'k');
colormap gray
for i = 1:N
  [~,yu_right,te,xe,ie] = ode113(@(t,x) fun_cr3bp(t,x,mu), tspan_u, XU_right(:,i), options_ODE_poincare);
  f1_p1 = plot(yu_right(:,1),yu_right(:,2),'Color','r');
  yu_right_f1(i,:) = xe(1,:);

  if size(xe,1) < 2
    continue;
  end
  yu_right_f2(i,:) = xe(2,:);
end
for i = 1:N
  [~,ys_left,te,xe,ie] = ode113(@(t,x) fun_cr3bp(t,x,mu), tspan_s, XS_left(:,i), options_ODE_poincare);
  ys_left = flipud(ys_left);
  f1_p2 = plot(ys_left(:,1),ys_left(:,2),'Color','#77AC30');
  ys_left_f1(i,:) = xe(1,:);

  if size(xe,1) < 2
    continue;
  end
  ys_left_f2(i,:) = xe(2,:);
end
plot(x_L1(:,1),x_L1(:,2),'k');
plot(x_L2(:,1),x_L2(:,2),'k');
plot(L1(1),L1(2),'*','MarkerFaceColor','k','MarkerEdgeColor','k','MarkerSize',15);
plot(L2(1),L2(2),'*','MarkerFaceColor','k','MarkerEdgeColor','k','MarkerSize',15);
xline(1-mu,'--k',"$x = 1 - \mu$",'Linewidth',1.5,'Fontsize',24,'Interpreter','latex');
view(gca,0,90);
axis equal
xlabel('$x$[-]');
ylabel('$y$[-]');
xlim([0.9 1.1]);
ylim([-0.08 0.08]);
grid on
legend([f1_p1, f1_p2], {'Unstable manifold','Stable manifold'});
hold off
f1_name = strcat('Ex8_poincare_map_traj_mu=',num2str(mu),'_xpert=',num2str(xpert));
f1_name = strrep(f1_name,'.',',');
save_fig(f1,f1_name,[0 90]);


f2 = figure();
hold on
f2_p1 = plot(yu_right_f1(:,2),yu_right_f1(:,5),'Color','r');
f2_p2 = plot(ys_left_f1(:,2),ys_left_f1(:,5),'Color','#77AC30');
f2_p3 = plot(yu_right_f2(:,2),yu_right_f2(:,5),'Color','m');
f2_p4 = plot(ys_left_f2(:,2),ys_left_f2(:,5),'Color','g');
axis square
xlabel('$y$[-]','Interpreter','latex');
ylabel('$\dot{y}$[-]','Interpreter','latex');
xlim([-0.03 0.05]);
ylim([-0.3 0.6]);
grid on
legend([f2_p1, f2_p2, f2_p3, f2_p4], {'1st insertion of unstable manifold','1st insertion of stable manifold',...
    '2nd insertion of unstable manifold','2nd insertion of stable manifold'},'Location','northeastoutside');
hold off
f2_name = strcat('Ex8_poincare_map_mu=',num2str(mu),'_xpert=',num2str(xpert));
f2_name = strrep(f2_name,'.',',');
save_fig(f2,f2_name,[0 90]);



%% End of script
time = strcat('calculation time: ', num2str(toc(myTimer)));
disp(time);

File: Example/Ex8_poincare_map.jl
"""
Ex8_poincare_map.jl
"""

using DifferentialEquations
using LinearAlgebra
using GLMakie
using Polynomials
using Printf
using CairoMakie
using Colors

current_dir = pwd()

# Function loading
functions_dir = normpath(joinpath(current_dir, "..", "Functions"))

if isdir(functions_dir)
    for file in readdir(functions_dir)
        if endswith(file, ".jl")
            full_path = joinpath(functions_dir, file)
            if isfile(full_path)
                include(full_path)
            else
                println("Warning: File not found - ", full_path)
            end
        end
    end
else
    println("Error: Functions directory not found at ", functions_dir)
end

# Start timer
t0 = time()  
# font setting
font = "Times New Roman"
# Save directory
save_fig_dir = "Figure/"

# Retrieving Parameters for the Sun-Jupiter Circular Restricted Three-Body Problem
mu, a_1, a_s, w_1 = fun_cr3bp_parameter(6)
# Calculation of Lagrange points
L1, L2, L3, L4, L5 = fun_libration_points(mu)

"""differential correction"""
iteration_DC_max = 100;
threshold = 1e-13;

# L1 Lyapunov orbit
x0_1 = [0.926672517344211 0 0  0 0.045684284381193 0]'
t0_1 = 2.915640890603734/2

for i = 1:iteration_DC_max
    global x0_1,t0_1,x_corrected
    global x_n_1,t_n_1,C_L1 = fun_differential_correction_cr3bp(x0_1, t0_1 ,mu)
    global prob_L1 = ODEProblem(fun_cr3bp!, x_n_1, (0.0, 2 * t_n_1), mu)
    global sol_L1 = solve(prob_L1, Vern7(), abstol=3e-14, reltol=1e-14)

    x_corrected = sol_L1.u

    # Calculate the difference between the initial state and the state after one orbital period.
    x_error = norm(x_corrected[end] .- x_corrected[1], 2)
    println(@sprintf("Iteration %d: x_error = %.4e", i, x_error))
    
    # The error has converged.
    if x_error < threshold
        break
    end

    # Stop the calculation if it diverges. 
    if x_error > 1e3
        println("Calculation diverged")
        return
    end

    # Within the maximum number of iterations, the error does not fall below the convergence threshold.
    if i == iteration_DC_max
        println("do not finish");
        return
    end
    
    # Update x0 and t0 
    x0_1 = x_n_1
    t0_1 = t_n_1
end

x_L1 = x_corrected
t_L1 = t_n_1

# L2 Lyapunov orbit
x0_2 = [1.072915577273646 0 0 0 -0.025945940908770 0]'
t0_2 = 3.184772418214781/2

for i in 1:iteration_DC_max
    global x0_2,t0_2,x_corrected
    global x_n_2,t_n_2,C_L2 = fun_differential_correction_cr3bp(x0_2, t0_2 ,mu)
    global prob_L2 = ODEProblem(fun_cr3bp!, x_n_2, (0.0, 2 * t_n_2), mu)
    global sol_L2 = solve(prob_L2, Vern7(), abstol=3e-14, reltol=1e-14)

    x_corrected = sol_L2.u

    # Calculate the difference between the initial state and the state after one orbital period.
    x_error = norm(x_corrected[end] .- x_corrected[1], 2)
    println(@sprintf("Iteration %d: x_error = %.4e", i, x_error))
    
    # The error has converged.
    if x_error < threshold
        break
    end

    # Stop the calculation if it diverges. 
    if x_error > 1e3
        println("Calculation diverged")
        return
    end

    # Within the maximum number of iterations, the error does not fall below the convergence threshold.
    if i == iteration_DC_max
        println("do not finish");
        return
    end
    
    # Update x0 and t0 
    x0_2 = x_n_2
    t0_2 = t_n_2
end

x_L2 = x_corrected
t_L2 = t_n_2

"""zero_velocity_curve"""
x_range = 0.9:1e-4:1.1
y_range = -0.1:1e-4:0.1
x = repeat(collect(x_range), 1, length(y_range)) 
y = repeat(collect(y_range)', length(x_range), 1)  
z = zeros(size(x))  

# Potential function calculation
r1 = sqrt.((x .+ mu) .^ 2 .+ y .^ 2 .+ z .^ 2)
r2 = sqrt.((x .- (1 - mu)) .^ 2 .+ y .^ 2 .+ z .^ 2)
U = 0.5 .* (x .^ 2 .+ y .^ 2) .+ (1 - mu) ./ r1 .+ mu ./ r2
C = 2 .* U

tspan_u = (0,9)
tspan_s = (10,0)
xpert = 1e-10
N = 40

# Initialize storage arrays:
yu_right_f1 = zeros(N, 6)
yu_right_f2 = fill(NaN, N, 6)
ys_left_f1  = zeros(N, 6)
ys_left_f2  = fill(NaN, N, 6)

# L1 unstable manifold generation
_, _, _, XU_right, _ = fun_manifold_cr3bp(mu, x_L1[1], 2 * t_L1, N, xpert)
# L2 stable manifold generation
XS_left, _, _, _, _ = fun_manifold_cr3bp(mu, x_L2[1], 2 * t_L2, N, xpert)

# Create a new figure and add a filled contour plot.
fig = Figure(size=(600, 600))

# Create an Axis for the Plot
ax = Axis(fig[1, 1]; aspect = 1, xlabel = "x[-]", ylabel = "y [-]",
        xlabelfont=font, ylabelfont=font,
        xticklabelfont=font, yticklabelfont=font
    )

# Create a filled contour at the level C_L1
GLMakie.contourf!(ax, x_range, y_range, C; levels=[minimum(C), C_L1], colormap=[:gray])
GLMakie.contour!(ax, x_range, y_range, C; levels = [C_L1], colormap = :grays)

# Right branch integration
for i in 1:N
    # Create an empty vector to store event states for the current integration
    local xe = Vector{Vector{Float64}}()

    # define a callback
    local cb_local = ContinuousCallback(condition, affect!(3, xe); save_positions=(false, false))
    
    # Define the ODE problem.
    local parameter = (mu)
    local prob = ODEProblem(fun_cr3bp!, XU_right[:, i], tspan_u, parameter)
    
    # Solve the ODE with high tolerances and the callback.
    local sol = solve(prob, Vern7(); reltol=3e-14, abstol=1e-14, callback=cb_local)
    # Reverse the trajectory (flip vertically) as in MATLAB's flipud.
    
    # Plot the reversed trajectory in color "#77AC30".
    GLMakie.lines!(sol[1,:], sol[2,:], color="red")
    
    # Save the event states.
    if length(xe) >= 1
        yu_right_f1[i, :] = xe[1]
    end
    if length(xe) >= 2
        yu_right_f2[i, :] = xe[2]
    end 
end

# --- Left branch integration ---
for i in 1:N
    local xe = Vector{Vector{Float64}}()

    local cb_local = ContinuousCallback(condition, affect!(3, xe); save_positions=(false, false))
    
    # Define the ODE problem.
    local parameter = (mu)
    local prob = ODEProblem(fun_cr3bp!, XS_left[:, i], tspan_s, parameter)
    
    # Solve the ODE with high tolerances and the callback.
    local sol = solve(prob, Vern7(); reltol=3e-14, abstol=1e-14, callback=cb_local)
    # Reverse the trajectory (flip vertically) as in MATLAB's flipud.
    
    # Plot the reversed trajectory in color "#77AC30".
    GLMakie.lines!(sol[1,:], sol[2,:], color="green")
    
    # Save the event states.
    if length(xe) >= 1
        ys_left_f1[i, :] = xe[1]
    end
    if length(xe) >= 2
        ys_left_f2[i, :] = xe[2]
    end
end

xlims!(ax, 0.9, 1.1)
ylims!(ax, -0.08, 0.08)

ax.xgridvisible[] = true
ax.ygridvisible[] = true

# Plot unstable manifold (x_L1) in black and assign a label.
p1 = GLMakie.lines!(ax, [v[1] for v in x_L1], [v[2] for v in x_L1], color = :black, label = "Unstable manifold")
# Plot stable manifold (x_L2) in black and assign a label.
p2 = GLMakie.lines!(ax, [v[1] for v in x_L2], [v[2] for v in x_L2], color = :black, label = "Stable manifold")

# Plot Lagrange points as star markers
GLMakie.scatter!(ax, [L1[1]], [L1[2]], markersize = 15, marker = :star5, color = :black)
GLMakie.scatter!(ax, [L2[1]], [L2[2]], markersize = 15, marker = :star5, color = :black)

p1 = GLMakie.lines!(ax, [v[1] for v in x_L1], [v[2] for v in x_L1], color = :black, label = "Unstable manifold")
# Plot stable manifold (x_L2) in black and assign a label.
p2 = GLMakie.lines!(ax, [v[1] for v in x_L2], [v[2] for v in x_L2], color = :black, label = "Stable manifold")

# Plot Lagrange points as star markers
GLMakie.scatter!(ax, [L1[1]], [L1[2]], markersize = 15, marker = :star5, color = :black)
GLMakie.scatter!(ax, [L2[1]], [L2[2]], markersize = 15, marker = :star5, color = :black)

# Add a vertical line at x = 1 - mu with a dashed line style.
GLMakie.lines!(ax, [1 - mu, 1 - mu], [-0.08, 0.08], linestyle = :dash, color = :black, linewidth = 1.5)

# Create a name for saving the figure
f1_name = "Ex8_poincare_map_traj_mu=$(mu)_xpert=$(xpert)"
f1_name = replace(f1_name, "." => ",")
# Display figure 1
display(fig)

# Poincare Map Plot
f2 = Figure(size = (800, 600))
ax2 = Axis(f2[1, 1], xlabel ="y[-]", ylabel ="y_dot[-]",
        xlabelfont=font, ylabelfont=font,
        xticklabelfont=font, yticklabelfont=font
    )

# Define a color for stable manifold insertion (hex color "#77AC30")
color_stable = parse(Colorant, "#77AC30")

# Plot the four insertion curves:
p2_1 = lines!(ax2, yu_right_f1[:, 2], yu_right_f1[:, 5], color = :red, label = "1st insertion of unstable manifold")
p2_2 = lines!(ax2, ys_left_f1[:, 2], ys_left_f1[:, 5], color = color_stable, label = "1st insertion of stable manifold")
p2_3 = lines!(ax2, yu_right_f2[:, 2], yu_right_f2[:, 5], color = :magenta, label = "2nd insertion of unstable manifold")
p2_4 = lines!(ax2, ys_left_f2[:, 2], ys_left_f2[:, 5], color = :green, label = "2nd insertion of stable manifold")

# Set axis square by matching data aspect ratio 
leg2 = Legend(f2, [p2_1, p2_2, p2_3, p2_4],
    ["1st insertion of unstable manifold", "1st insertion of stable manifold",
     "2nd insertion of unstable manifold", "2nd insertion of stable manifold"],
    tellwidth = false, markersize = 10, labelfont = font
)
# You can adjust the legend position by modifying layout; here we add it to the right of the axis.
f2[1, 2] = leg2

# Create a name for saving the figure.
f2_name = "Ex8_poincare_map_mu=$(mu)_xpert=$(xpert)"
f2_name = replace(f2_name, "." => ",")

# Display figure 2
display(f2)